### DISCUSSION TITLE
Consider using yagfuncdata

### DISCUSSION THREAD

**Discussion by jo3-l** [l-zeuch/yagpdb.vim#64] [Ideas]
https://github.com/jo3-l/yagfuncdata is a little CLI tool I wrote to automatically fetch names of YAGPDB template functions. From the README:
> yagfuncdata aims to simplify workflows of projects that require a list of YAGPDB template functions, such as language support extensions for editors. For example, one could imagine setting up a cronjob in CI to run yagfuncdata regularly to check for updates and regenerate some asset file if anything changed.
>
> Note, however, that it is not recommended to run yagfuncdata directly as part of an application as the command is not infallible. Changes to the structure of relevant files defining template functions may cause the generated list to be incomplete or downright wrong. (We elaborate on why this is the case below.) Thus, its output should be manually vetted by a human against a reliable baseline.

Perhaps `yagfuncdata` could be of use to this project, which seems to maintain a similar list as part of the syntax definition that requires manual updating.

**Comment by l-zeuch** [comment]
Thanks for this!

As per our private conversation, I am currently working on something else, but will take some time to think about a viable way of incorporating this.

As it stands, we'd need to write another script to parse the bare-boned output of the CLI tool to incorporate into our syntax definitions, whilst keeping the format our `gen_code_completion.py` tool  looks for and matches against in mind. However, I don't think this should be a too hard of an undertaking.

The main issue I'm currently seeing is the unstable output given by upstream; *if* we implement some sort of automated bump, it should be completely stable and reliable, otherwise it is hardly an upgrade from the current status quo: updating the syntax file(s) manually.

---
Meanwhile: cc @LRitzdorf

**Comment by jo3-l** [comment]
> The main issue I'm currently seeing is the unstable output given by upstream; if we implement some sort of automated bump, it should be completely stable and reliable, otherwise it is hardly an upgrade from the current status quo: updating the syntax file(s) manually.

I had hoped to address this in my original comment:

> For example, one could imagine setting up a cronjob in CI to run yagfuncdata regularly to check for updates and regenerate some asset file if anything changed.

That is, the intended workflow is to run `genfuncdata` regularly to generate the necessary files. Then, issue a PR if any changes occurred (which should then be manually reviewed.) This obviates any concerns relating to unstable or possibly broken output while still gaining much benefit from the added automation.

**Reply by l-zeuch** [reply]
Yeah, I was just reiterating on this.

If done properly, it would be similar to Dependabot :thinking:

**Reply by LRitzdorf** [reply]
Ooh, I definitely like this concept!

**Reply by jo3-l** [reply]
FWIW, I set up something similar to what I described in the OP for `action-check-yag-tmpl-syntax` today; perhaps it can be of some use. See the [workflow](https://github.com/jo3-l/action-check-yag-tmpl-syntax/blob/main/.github/workflows/regenerate-funcs.yml), an [example PR](https://github.com/jo3-l/action-check-yag-tmpl-syntax/pull/6), and the [code generator](https://github.com/jo3-l/action-check-yag-tmpl-syntax/blob/main/gen/gen_funcs.go). (Note, however, that `action-check-yag-tmpl-syntax` uses the Go API that `yagfuncdata` provides as opposed to the command-line tool, so the setup will not be the exact same.)

**Reply by l-zeuch** [reply]
Would you be willing to write a patchset for this? If so, we're happy to review it.

Otherwise, no problem, we can probably write one ourselves.

----
**Edit:**
The main problem here is parsing the output to a vim syntax file, and also parsing the output to our current completion implementation.

**Reply by jo3-l** [reply]
Unfortunately I do not think I am sufficiently familiar with the internals of this project to write a PR. That said, I am of course happy to assist with usage of the `yagfuncdata` tool and setting up the GitHub workflow -- just not the code generator itself.
